### ISSUE TITLE
Step 18: Integrated Deletion-Pattern Piecewise Regression & Dynamic Nonlinear Breakpoint Analysis Pipeline (Stage 3 Multi-System Transition Engine)

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/JapanLongevityAI#22]
# ============================================================
# Generic deletion-pattern breakpoint analysis
# Clustering + high-risk cluster selection + breakpoint analysis
# patient_TT 20260511.csv
# ============================================================

from google.colab import drive
drive.mount('/content/drive')

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from scipy.stats import spearmanr
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.decomposition import PCA
from statsmodels.nonparametric.smoothers_lowess import lowess

# ============================================================
# 1. Settings
# ============================================================

BASE = "/content/drive/MyDrive"
FILE_PATH = f"{BASE}/patient_TT 20260511.csv"

OUT_DIR = f"{BASE}/deletion_breakpoint_analysis_20260514"
os.makedirs(OUT_DIR, exist_ok=True)

# Change or add markers here
DROP_MARKERS = ["TT", "Cre", "CRP"]

# Add BUN if present in your dataset
# DROP_MARKERS = ["TT", "Cre", "CRP", "BUN"]

BREAKPOINT_MARKERS = ["TT", "CRP", "Cre", "Alb", "Hb", "BMI"]

BASE_FEATURES = [
    "age", "height", "BW", "BMI",
    "Hb", "Hct", "Alb",
    "AST", "ALT", "ALP",
    "Cre", "Glu", "TG", "HDL-C", "LDL-C",
    "LH", "FSH", "CRP", "TT", "BUN"
]

RISK_POSITIVE = [
    "age", "CRP", "Cre", "BUN",
    "Glu", "TG", "AST", "ALT", "ALP"
]

RISK_NEGATIVE = [
    "Alb", "Hb", "Hct", "HDL-C", "TT", "BMI"
]

# ============================================================
# 2. Load data
# ============================================================

df_raw = pd.read_csv(
    FILE_PATH,
    na_values=["NA", "NaN", "#VALUE!", "", " "]
)

print("Original shape:", df_raw.shape)
display(df_raw.head())

for col in df_raw.columns:
    df_raw[col] = pd.to_numeric(df_raw[col], errors="coerce")

# ============================================================
# 3. Helper functions
# ============================================================

def fit_piecewise_grid(data, x_col, y_col, min_age=40, max_age=90):
    """
    One-breakpoint piecewise linear regression by grid search.
    y = b0 + b1*x + b2*max(0, x - breakpoint)
    Breakpoint selected by minimum RSS.
    """

    tmp = data[[x_col, y_col]].dropna().copy()
    tmp = tmp.replace([np.inf, -np.inf], np.nan).dropna()

    if len(tmp) < 50:
        return None

    x = tmp[x_col].values.astype(float)
    y = tmp[y_col].values.astype(float)

    best = None

    for bp in range(min_age, max_age + 1):

        X = np.column_stack([
            np.ones_like(x),
            x,
            np.maximum(0, x - bp)
        ])

        beta = np.linalg.lstsq(X, y, rcond=None)[0]
        yhat = X @ beta
        rss = np.sum((y - yhat) ** 2)

        if best is None or rss < best["rss"]:
            best = {
                "breakpoint_age": bp,
                "rss": rss,
                "intercept": beta[0],
                "slope_before": beta[1],
                "slope_after": beta[1] + beta[2],
                "n": len(tmp)
            }

    return best

def run_deletion_clustering(df_raw, drop_marker, out_dir):
    """
    Runs deletion-pattern clustering for one marker.
    Returns:
        df_result, cluster_risk, cluster_z, high_risk_cluster, feature_cols
    """

    condition = f"no_{drop_marker}"

    print("\n================================================")
    print(f"Running deletion condition: {condition}")
    print("================================================")

    df_model = df_raw.drop(columns=[drop_marker], errors="ignore").copy()

    for col in df_model.columns:
        df_model[col] = pd.to_numeric(df_model[col], errors="coerce")

    feature_cols = [
        c for c in BASE_FEATURES
        if c in df_model.columns
    ]

    df_model = df_model[feature_cols].copy()

    print("Model shape:", df_model.shape)
    print("Dropped marker:", drop_marker)

    missing_rate = df_model.isna().mean().sort_values(ascending=False)
    print("Missing rate:")
    display(missing_rate)

    imputer = SimpleImputer(strategy="median")
    X_imputed = imputer.fit_transform(df_model)

    df_clean = pd.DataFrame(
        X_imputed,
        columns=feature_cols
    )

    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(df_clean)

    sil_rows = []

    for k in range(2, 8):
        model = KMeans(
            n_clusters=k,
            random_state=0,
            n_init=50
        )
        labels = model.fit_predict(X_scaled)
        score = silhouette_score(X_scaled, labels)

        sil_rows.append({
            "condition": condition,
            "k": k,
            "silhouette": score
        })

    sil_df = pd.DataFrame(sil_rows)

    best_k = int(
        sil_df.loc[
            sil_df["silhouette"].idxmax(),
            "k"
        ]
    )

    print("Silhouette results:")
    display(sil_df)

    print("Best k:", best_k)

    kmeans = KMeans(
        n_clusters=best_k,
        random_state=0,
        n_init=100
    )

    labels = kmeans.fit_predict(X_scaled)

    cluster_col = f"cluster_{condition}"

    df_result = df_clean.copy()
    df_result[cluster_col] = labels

    scaled_df = pd.DataFrame(
        X_scaled,
        columns=feature_cols
    )
    scaled_df[cluster_col] = labels

    risk_score_col = f"aging_risk_score_{condition}"

    df_result[risk_score_col] = 0.0

    available_pos = [
        v for v in RISK_POSITIVE
        if v in feature_cols
    ]

    available_neg = [
        v for v in RISK_NEGATIVE
        if v in feature_cols
    ]

    for v in available_pos:
        df_result[risk_score_col] += scaled_df[v]

    for v in available_neg:
        df_result[risk_score_col] -= scaled_df[v]

    agg_dict = {
        "n": (cluster_col, "size"),
        "mean_age": ("age", "mean"),
        f"mean_{risk_score_col}": (risk_score_col, "mean"),
        f"median_{risk_score_col}": (risk_score_col, "median")
    }

    for marker in ["TT", "CRP", "Cre", "BUN", "Alb", "Hb", "BMI"]:
        if marker in df_result.columns:
            agg_dict[f"mean_{marker}"] = (marker, "mean")

    cluster_risk = (
        df_result
        .groupby(cluster_col)
        .agg(**agg_dict)
        .sort_values(f"mean_{risk_score_col}", ascending=False)
    )

    print("Cluster counts:")
    display(df_result[cluster_col].value_counts().sort_index())

    print("Risk-ranked clusters:")
    display(cluster_risk)

    # Exclude singleton clusters for high-risk interpretation
    non_singleton = cluster_risk[cluster_risk["n"] > 1].copy()

    if non_singleton.empty:
        high_risk_cluster = cluster_risk.index[0]
    else:
        high_risk_cluster = non_singleton.index[0]

    print("Selected high-risk cluster:", high_risk_cluster)

    cluster_z = scaled_df.groupby(cluster_col).mean()

    # PCA
    pca = PCA(n_components=2, random_state=0)
    X_pca = pca.fit_transform(X_scaled)

    df_result["PC1"] = X_pca[:, 0]
    df_result["PC2"] = X_pca[:, 1]

    # Save
    condition_dir = f"{out_dir}/{condition}"
    os.makedirs(condition_dir, exist_ok=True)

    df_result.to_csv(
        f"{condition_dir}/{condition}_cluster_result.csv",
        index=False,
        encoding="utf-8-sig"
    )

    cluster_risk.to_csv(
        f"{condition_dir}/{condition}_cluster_risk_summary.csv",
        encoding="utf-8-sig"
    )

    cluster_z.to_csv(
        f"{condition_dir}/{condition}_cluster_zprofile.csv",
        encoding="utf-8-sig"
    )

    sil_df.to_csv(
        f"{condition_dir}/{condition}_silhouette_results.csv",
        index=False,
        encoding="utf-8-sig"
    )

    # PCA figure
    plt.figure(figsize=(7, 6))

    for c in sorted(df_result[cluster_col].unique()):
        tmp = df_result[df_result[cluster_col] == c]
        plt.scatter(
            tmp["PC1"],
            tmp["PC2"],
            label=f"Cluster {c}",
            alpha=0.55,
            s=12
        )

    plt.xlabel("PC1")
    plt.ylabel("PC2")
    plt.title(f"Clustering under {condition}")
    plt.legend()
    plt.grid(alpha=0.25)
    plt.tight_layout()

    plt.savefig(
        f"{condition_dir}/{condition}_PCA_clusters.png",
        dpi=600,
        bbox_inches="tight"
    )

    plt.show()

    # z-profile figure
    plt.figure(figsize=(12, 5))
    plt.imshow(cluster_z, aspect="auto")
    plt.colorbar(label="Mean z-score")
    plt.xticks(
        range(len(cluster_z.columns)),
        cluster_z.columns,
        rotation=90
    )
    plt.yticks(
        range(len(cluster_z.index)),
        cluster_z.index
    )
    plt.title(f"Cluster biomarker profile under {condition}")
    plt.tight_layout()

    plt.savefig(
        f"{condition_dir}/{condition}_zprofile_heatmap.png",
        dpi=600,
        bbox_inches="tight"
    )

    plt.show()

    return {
        "condition": condition,
        "drop_marker": drop_marker,
        "df_result": df_result,
        "cluster_col": cluster_col,
        "high_risk_cluster": high_risk_cluster,
        "cluster_risk": cluster_risk,
        "cluster_z": cluster_z,
        "silhouette": sil_df,
        "feature_cols": feature_cols
    }

def run_breakpoint_analysis(result_obj, out_dir):
    """
    Runs age-marker correlations and piecewise regression
    for overall cohort and selected high-risk cluster.
    """

    condition = result_obj["condition"]
    df_result = result_obj["df_result"]
    cluster_col = result_obj["cluster_col"]
    high_risk_cluster = result_obj["high_risk_cluster"]

    condition_dir = f"{out_dir}/{condition}"

    df_high = df_result[
        df_result[cluster_col] == high_risk_cluster
    ].copy()

    scopes = {
        "overall": df_result,
        "high_risk_cluster": df_high
    }

    corr_rows = []
    break_rows = []

    target_markers = [
        m for m in BREAKPOINT_MARKERS
        if m in df_result.columns
    ]

    for scope_name, data in scopes.items():

        for marker in target_markers:

            tmp = data[["age", marker]].dropna().copy()

            if len(tmp) >= 10:
                rho, pval = spearmanr(tmp["age"], tmp[marker])
            else:
                rho, pval = np.nan, np.nan

            corr_rows.append({
                "condition": condition,
                "drop_marker": result_obj["drop_marker"],
                "scope": scope_name,
                "marker": marker,
                "n": len(tmp),
                "spearman_r_age": rho,
                "spearman_p_age": pval
            })

            res = fit_piecewise_grid(
                data=data,
                x_col="age",
                y_col=marker,
                min_age=40,
                max_age=90
            )

            if res is not None:
                res.update({
                    "condition": condition,
                    "drop_marker": result_obj["drop_marker"],
                    "scope": scope_name,
                    "marker": marker
                })
                break_rows.append(res)

    corr_df = pd.DataFrame(corr_rows)
    break_df = pd.DataFrame(break_rows)

    corr_df.to_csv(
        f"{condition_dir}/{condition}_spearman_age_marker.csv",
        index=False,
        encoding="utf-8-sig"
    )

    break_df.to_csv(
        f"{condition_dir}/{condition}_piecewise_breakpoints.csv",
        index=False,
        encoding="utf-8-sig"
    )

    display(corr_df)
    display(break_df)

    return corr_df, break_df

def plot_breakpoint_panels(result_obj, break_df, out_dir):
    """
    Creates 2-row figure:
        top row = overall cohort
        bottom row = selected high-risk cluster
    """

    condition = result_obj["condition"]
    df_result = result_obj["df_result"]
    cluster_col = result_obj["cluster_col"]
    high_risk_cluster = result_obj["high_risk_cluster"]

    condition_dir = f"{out_dir}/{condition}"

    df_high = df_result[
        df_result[cluster_col] == high_risk_cluster
    ].copy()

    target_markers = [
        m for m in BREAKPOINT_MARKERS
        if m in df_result.columns
    ]

    n_markers = len(target_markers)

    fig, axes = plt.subplots(
        2,
        n_markers,
        figsize=(4.8 * n_markers, 8.5),
        squeeze=False
    )

    for row_idx, (scope_name, data) in enumerate([
        ("overall", df_result),
        ("high_risk_cluster", df_high)
    ]):

        for col_idx, marker in enumerate(target_markers):

            ax = axes[row_idx, col_idx]

            tmp = data[["age", marker]].dropna().copy()

            ax.scatter(
                tmp["age"],
                tmp[marker],
                s=12,
                alpha=0.25
            )

            if len(tmp) >= 30:
                lo = lowess(
                    tmp[marker],
                    tmp["age"],
                    frac=0.30,
                    return_sorted=True
                )

                ax.plot(
                    lo[:, 0],
                    lo[:, 1],
                    linewidth=2.4
                )

            bp_row = break_df[
                (break_df["scope"] == scope_name) &
                (break_df["marker"] == marker)
            ]

            if len(bp_row) > 0:
                bp = float(bp_row.iloc[0]["breakpoint_age"])
                ax.axvline(
                    bp,
                    linestyle="--",
                    linewidth=1.4
                )

            ax.set_title(f"{scope_name}: age vs {marker}")
            ax.set_xlabel("Age (years)")
            ax.set_ylabel(marker)
            ax.grid(alpha=0.25)

    plt.suptitle(
        f"Age-related breakpoint patterns under {condition}",
        fontsize=16,
        fontweight="bold"
    )

    plt.tight_layout(rect=[0, 0, 1, 0.95])

    fig_path_png = f"{condition_dir}/{condition}_breakpoint_panels.png"
    fig_path_pdf = f"{condition_dir}/{condition}_breakpoint_panels.pdf"

    plt.savefig(fig_path_png, dpi=600, bbox_inches="tight")
    plt.savefig(fig_path_pdf, bbox_inches="tight")
    plt.show()

    print("Saved:")
    print(fig_path_png)
    print(fig_path_pdf)

# ============================================================
# 4. Run all deletion conditions
# ============================================================

all_cluster_results = []
all_corr = []
all_breaks = []

for drop_marker in DROP_MARKERS:

    if drop_marker not in df_raw.columns:
        print(f"Skipping {drop_marker}: not found in dataset.")
        continue

    result_obj = run_deletion_clustering(
        df_raw=df_raw,
        drop_marker=drop_marker,
        out_dir=OUT_DIR
    )

    corr_df, break_df = run_breakpoint_analysis(
        result_obj=result_obj,
        out_dir=OUT_DIR
    )

    plot_breakpoint_panels(
        result_obj=result_obj,
        break_df=break_df,
        out_dir=OUT_DIR
    )

    all_cluster_results.append(result_obj)

    all_corr.append(corr_df)
    all_breaks.append(break_df)

# ============================================================
# 5. Integrated summary outputs
# ============================================================

all_corr_df = pd.concat(all_corr, ignore_index=True)
all_break_df = pd.concat(all_breaks, ignore_index=True)

all_corr_path = f"{OUT_DIR}/ALL_deletion_spearman_age_marker.csv"
all_break_path = f"{OUT_DIR}/ALL_deletion_piecewise_breakpoints.csv"

all_corr_df.to_csv(
    all_corr_path,
    index=False,
    encoding="utf-8-sig"
)

all_break_df.to_csv(
    all_break_path,
    index=False,
    encoding="utf-8-sig"
)

display(all_corr_df)
display(all_break_df)

print("Saved:")
print(all_corr_path)
print(all_break_path)

# ============================================================
# 6. Breakpoint heatmap table
# ============================================================

heatmap_df = (
    all_break_df
    .pivot_table(
        index=["condition", "scope"],
        columns="marker",
        values="breakpoint_age",
        aggfunc="first"
    )
    .reset_index()
)

heatmap_path = f"{OUT_DIR}/breakpoint_age_heatmap_table.csv"

heatmap_df.to_csv(
    heatmap_path,
    index=False,
    encoding="utf-8-sig"
)

display(heatmap_df)

print("Saved:")
print(heatmap_path)

# ============================================================
# 7. Breakpoint heatmap figure
# ============================================================

plot_df = heatmap_df.copy()
plot_df["row_label"] = plot_df["condition"] + " / " + plot_df["scope"]

marker_cols = [
    c for c in BREAKPOINT_MARKERS
    if c in plot_df.columns
]

mat = plot_df[marker_cols].values.astype(float)

plt.figure(figsize=(1.4 * len(marker_cols) + 6, 0.55 * len(plot_df) + 4))

plt.imshow(
    mat,
    aspect="auto"
)

plt.colorbar(label="Breakpoint age")

plt.xticks(
    range(len(marker_cols)),
    marker_cols,
    rotation=45,
    ha="right"
)

plt.yticks(
    range(len(plot_df)),
    plot_df["row_label"]
)

for i in range(mat.shape[0]):
    for j in range(mat.shape[1]):
        if not np.isnan(mat[i, j]):
            plt.text(
                j,
                i,
                f"{mat[i, j]:.0f}",
                ha="center",
                va="center",
                fontsize=9
            )

plt.title("Breakpoint age under biomarker deletion conditions")
plt.tight_layout()

heatmap_png = f"{OUT_DIR}/Figure_breakpoint_age_heatmap.png"
heatmap_pdf = f"{OUT_DIR}/Figure_breakpoint_age_heatmap.pdf"

plt.savefig(heatmap_png, dpi=600, bbox_inches="tight")
plt.savefig(heatmap_pdf, bbox_inches="tight")
plt.show()

print("Saved:")
print(heatmap_png)
print(heatmap_pdf)

# ============================================================
# 8. Cluster summary table
# ============================================================

cluster_summary_rows = []

for obj in all_cluster_results:

    condition = obj["condition"]
    drop_marker = obj["drop_marker"]
    high_cluster = obj["high_risk_cluster"]
    cluster_risk = obj["cluster_risk"].copy()

    row = cluster_risk.loc[high_cluster].to_dict()

    row_out = {
        "condition": condition,
        "drop_marker": drop_marker,
        "selected_high_risk_cluster": high_cluster,
    }

    row_out.update(row)

    cluster_summary_rows.append(row_out)

cluster_summary_df = pd.DataFrame(cluster_summary_rows)

cluster_summary_path = f"{OUT_DIR}/ALL_deletion_high_risk_cluster_summary.csv"

cluster_summary_df.to_csv(
    cluster_summary_path,
    index=False,
    encoding="utf-8-sig"
)

display(cluster_summary_df)

print("Saved:")
print(cluster_summary_path)
